src/relay/pass/gradient.cc

```diff
@@ -279,29 +279,88 @@ struct ReverseAD : ExprMutator {
           args.push_back(ll->Push(VisitExpr(arg)));
         }
         std::vector<Expr> orig_args;
-        for (const auto& arg : args) {
-          orig_args.push_back(GetField(arg, 0));
+        std::function<Expr(const Type& t, const Expr& e)> get_orig_args;
+        get_orig_args = [&](const Type& t, const Expr& e) -> Expr{
```

Comment:
Is there a reason this, `get_grad`, and `get_ret` should be local lambdas and not factored out? It particularly seems like it might be worth factoring these out somewhat since they all only differ in the tensor type case. It makes it somewhat difficult to understand what's going on here (perhaps more comments would help)